### ISSUE TITLE
getCumulativeBytes return number is smaller than Peak Memory on Prestissimo

### ISSUE THREAD

**Issue by GuChangyang** [facebookincubator/velox#3456]
### Description

I call tracker->getCumulativeBytes() in prestimisso codebase : https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoTask.cpp#L240

and then I found the method getCumulativeBytes return a number which is smaller than Peak Memory, is this as expected?

I am thinking CumulativeMemory is total memory and it should be larger than peak memory. am I right?

Thanks

**Comment by kagamiori** [comment]
cc @mbasmanova

**Comment by GuChangyang** [comment]
Follow-up question : 

"prestoTaskStats.cumulativeUserMemory = {};" on PrestoTask.cpp is intended null
is it because the tracker collects cumulativeUserMemory which is incomplete?

I think the change is very simple , just changing to below

prestoTaskStats.cumulativeUserMemory = tracker->getCumulativeBytes();

will above change track all users memory?

Thank you!

**Comment by mbasmanova** [comment]
@xiaoxmeng @spershin Sergey, Meng, would you help answer these questions?

**Comment by xiaoxmeng** [comment]
@GuChangyang yes, getCumulativeBytes accumulate the memory usage and is supposed to be larger than the peak. But if you make two calls and get accumulate bytes in the first call and peak in second. It might have in some race condition. Are you always seeing this inconsistency? I am refactoring this part of code. Also can you show me the diff that how you collect those stats? Thanks!

**Comment by GuChangyang** [comment]
Yes, I always see peak memory is larger than the number from getCumulativeBytes,

I just change the code to "prestoTaskStats.cumulativeUserMemory = tracker->getCumulativeBytes();" and then deploy it on server and I can see the number from the presto UI.

Also I checked the java side prestodb codebase : https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java#L558-571

It uses cumulativeBytes * execution time to get the cumulativeUserMemory.

Curious do we still need multiply the execution time cpp codebase? thank you

**Comment by GuChangyang** [comment]
Any new comments here thank you!

**Comment by xiaoxmeng** [comment]
Can you point me the code that collect the stats? Is it submitted to Presto native code base? It seems to be a bug from what you have described. The accumulative bytes in Velox is always accumulated and never averaged out with time. Thanks!

**Comment by markjin1990** [comment]
> https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java#L558-571

@xiaoxmeng Please see above.

**Comment by xiaoxmeng** [comment]
Per offline discussion with @markjin1990 @GuChangyang: it looks like the Presto java and c++ has different handling on the accumulative bytes. The java side code: [link](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java#L562) , and cpp side code: [link](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoTask.cpp#L240). Java side collect the accumulative bytes based on the averaged current memory usage (a sort of a memory usage at a snapshot) * time period in milliseconds, and then send it to Coordinator for further processing. Velox's memory tracker collects the accumulative bytes as a monotonically increasing bytes count so which means we could do a more accurate calculation than Java side. Presto CPP can record the last accumulative bytes, current accumulative bytes - last accumulative bytes =. accumulative memory allocations between two collections. Otherwise, we can do the same calculation as presto java side. I am not sure what's the exact semantics here we want at Presto coordinator side. @mbasmanova WDYT? Thanks!

**Comment by xiaoxmeng** [comment]
Have discuss this with @mbasmanova
```
For peak memory usage, we have these fields:

prestoTaskStats.peakUserMemoryInBytes = tracker->getPeakUserBytes();
prestoTaskStats.peakTotalMemoryInBytes = tracker->getPeakTotalBytes();

cumulative usage is something else:

            cumulativeUserMemory.addAndGet(averageUserMemoryForLastPeriod * sinceLastPeriodMillis);
            cumulativeTotalMemory.addAndGet(averageTotalMemoryForLastPeriod * sinceLastPeriodMillis);

I think this is memory usage times duration of usage. It is used to distinguish between a query that uses 10GB for 10 min vs. a query that uses 100GB for 2 seconds.

First query uses a total of 10 * 600 = 6K byte-seconds, while the second query uses only 100 * 2 = 200 byte seconds.

The second query needs "less" memory than the first even though peak memory for the 2nd query is a lot higher.
```
Given that, we can't use the accumulative bytes in tracker from Velox and the accumulative here is "memory usage times duration of usage.". So we need to make this stats collection consistent with Java side.

**Comment by GuChangyang** [comment]
@xiaoxmeng, Thanks for the updates, I will make some changes based on above and submit a pr lately.

**Comment by GuChangyang** [comment]
@xiaoxmeng @markjin1990  I submitted a review request : https://github.com/prestodb/presto/pull/18991
please take a look.

**Comment by GuChangyang** [comment]
@xiaoxmeng I noticed you refactored MemoryTrackerUsage file and it  is only recording the total memory and no more user memory, and also submitted this MR to match the new refactor code : https://github.com/prestodb/presto/pull/18991
